dill/_dill.py

```diff
@@ -504,29 +505,138 @@ def dump_session(filename='/tmp/session.pkl', main=None, byref=False, **kwds):
+        if isinstance(error, AttributeError) and not main_is_none:
+            # File is not peekable, but we have main.
+            return None
+        raise UnpicklingError("unable to identify main module") from error
+
+def load_session(filename='/tmp/session.pkl', main=None, **kwds):
+    """update the __main__ module with the state from the session file"""
+    file = _open_peekable(filename)
+    try:
+        #FIXME: dill.settings are disabled
+        unpickler = Unpickler(file, **kwds)
         unpickler._session = True
+        pickle_main = _inspect_pickle(file, main is None)
+
+        # Resolve unpickler._main
+        if main is None and pickle_main is not None:
+            main = pickle_main
+        if isinstance(main, str):
+            if main.startswith('__runtime__.'):
+                # Create runtime module to load the session into.
+                main = ModuleType(main.partition('.')[-1])
+            else:
+                main = _import_module(main)
+        if main is not None:
+            if not isinstance(main, ModuleType):
+                raise ValueError("%r is not a module" % main)
+            unpickler._main = main
+        main = unpickler._main
+
+        # Check against the pickle's main.
+        is_main_imported = _is_imported_module(main)
+        if pickle_main is not None:
+            is_runtime_mod = pickle_main.startswith('__runtime__.')
+            if is_runtime_mod:
+                pickle_main = pickle_main.partition('.')[-1]
+            if is_runtime_mod and is_main_imported:
+                raise UnpicklingError("can't restore non-imported module %r into an imported one" \
+                                      % pickle_main)
+            if not is_runtime_mod and not is_main_imported:
+                raise UnpicklingError("can't restore imported module %r into a non-imported one" \
+                                      % pickle_main)
+            if main.__name__ != pickle_main:
+                raise UnpicklingError("can't restore module %r into module %r" \
+                                      % (pickle_main, main.__name__))
+
+        # This is for find_class() to be able to locate it.
+        if not is_main_imported:
+            runtime_main = '__runtime__.%s' % main.__name__
+            sys.modules[runtime_main] = main
+
         module = unpickler.load()
-        unpickler._session = False
-        main.__dict__.update(module.__dict__)
-        _restore_modules(unpickler, main)
     finally:
-        if f is not filename:  # If newly opened file
-            f.close()
-    return
+        if not hasattr(filename, 'read'):  # if newly opened file
+            file.close()
+        try:
+            del sys.modules[runtime_main]
+        except (KeyError, NameError):
+            pass
+    assert module is main
+    _restore_modules(unpickler, module)
+    if module is not _main_module:
+        return module
+
+def load_session_copy(filename='/tmp/session.pkl', **kwds):
```

Comment:
I don't like the name of this function.

Reply:
Maybe a better choice is to rename it `_load_session`, then have it called seamlessly from `load_session` if someone sets a `kwarg` (where the default `None` would be `__main__`).